lib/azure/armrest/resource_group_based_service.rb

```diff
@@ -35,19 +66,34 @@ def list(rgroup = configuration.resource_group)
       def list_all(filter = {})
         url = build_url
         url = yield(url) || url if block_given?
+
         response = rest_get(url)
-        results = JSON.parse(response)['value'].map { |hash| model_class.new(hash) }
+        json_response = JSON.parse(response)['value']
+
+        results = ArmrestCollection.new(json_response.map { |hash| model_class.new(hash) })
+
         filter.empty? ? results : results.select { |obj| filter.all? { |k, v| obj.public_send(k) == v } }
+
+        results.response_headers = Azure::Armrest::ResponseHeaders.new(response.headers)
```

Comment:
You should use new `results` from above line. But it is no longer needed if you refactor the initializer as suggested.